### PR TITLE
CP-25630: Change QEMU startup

### PR DESCRIPTION
**Pull request by rosslagerwall**

### CODE REVIEW

**Comment by rosslagerwall** [comment]
Note that there is a corresponding QEMU PR to avoid writing to xenstore.

I have dev-tested startup and migration for both QEMU and qemu-trad.

File: xc/device.ml [hunk 1]
```diff
@@ -1800,39 +1798,45 @@ module Dm_Common = struct
...
-    debug "Daemon initialised: %s" syslog_key;
+    debug "Daemon started: %s" syslog_key;
     pid
 
+  (* Waits for a daemon to signal startup by writing to a xenstore path
+   * (optionally with a given value) If this doesn't happen in the timeout then
+   * an exception is raised *)
+  let wait_path ~pid ~task ~name ~domid ~xs ~ready_path ?ready_val ~timeout
+      ~cancel _ =
+    let syslog_key = (Printf.sprintf "%s-%d" name domid) in
```
Comment (lindig) [inline xc/device.ml]: `(` and `)` are not required.

File: xc/device.ml [hunk 2]
```diff
@@ -1800,39 +1798,45 @@ module Dm_Common = struct
...
+    let finished = ref false in
+    let watch = Watch.value_to_appear ready_path |> Watch.map (fun _ -> ()) in
+    let start = Mtime_clock.counter () in
+    while (Mtime.Span.to_ms (Mtime_clock.count start)) < timeout && not !finished do
```
Comment (lindig) [inline xc/device.ml]: A slightly more idiomatic way to check for a timeout is described here: 
https://discuss.ocaml.org/t/monotonic-clock-mtime/923/2

File: xc/device.ml [hunk 3]
```diff
@@ -1800,39 +1798,45 @@ module Dm_Common = struct
...
+      Xenops_task.check_cancelling task;
+      try
+        let (_: bool) = cancellable_watch cancel [ watch ] [] task ~xs ~timeout () in
+        let state = try xs.Xs.read ready_path with _ -> "" in
+        match ready_val with
+        | Some value ->
```
Comment (lindig) [inline xc/device.ml]: This is a matter of taste, you could do:
```
match ready_val with
| Some value when value = state -> finished := true
| Some _ -> raise ..
| None -> finished := false
```

File: xc/device.ml [hunk 4]
```diff
@@ -2185,8 +2196,23 @@ module Backend = struct
                   Qmp_protocol.write c (Command (None, Remove_fd fd_info.fdset_id))))
           (fun () -> Qmp_protocol.close c)
 
+      (* Wait for QEMU's event socket to appear. *)
+      let wait_event_socket ~task ~name ~domid ~timeout =
+        let finished = ref false in
+        let start = Mtime_clock.counter () in
+        while (Mtime.Span.to_ms (Mtime_clock.count start)) < timeout && not !finished do
```
Comment (lindig) [inline xc/device.ml]: See above for checking a timeout.

File: xc/device.ml (hunk 3 above)
Comment (rosslagerwall) [reply xc/device.ml]: Note that apart from changing to a monotonic clock, the wait_path function is split out from the original init_daemon but I didn't actually change any of the code. I'm not sure if GitHub has the ability to diff ignoring whitespace changes, but that would be useful.

File: xc/device.ml (hunk 2 above)
Comment (rosslagerwall) [reply xc/device.ml]: OK, I'll take a look at this. I just copied what the code was doing previously but using gettimeofday().

File: xc/device.ml (hunk 3 above)
Comment (mseri) [reply xc/device.ml]: You can add ?w=false to the url. Good to look at the changes but you cannot comment in ignore-whitespace mode

**Comment by rosslagerwall** [comment]
The extra commits should address the review comments.

**Comment by lindig** [comment]
Maybe @mseri can have a look, too. I'd be ready to merge this.

File: scripts/qemu-wrapper [hunk 5]
```diff
@@ -382,6 +382,7 @@ def main(argv):
         qemu_args += ["-runas", "%d.%d" % (uid, gid)]
 
     qemu_args += ["-sandbox", "on,obsolete=deny,elevateprivileges=allow,spawn=deny,resourcecontrol=deny"]
+    qemu_args += ["-S"]
```
Comment (mseri) [inline scripts/qemu-wrapper]: Just as a personal curiosity, why we now want to prevent the CPU to start?

File: xc/device.ml [hunk 6]
```diff
@@ -2073,6 +2083,7 @@ module Backend = struct
             Lookup.channel_of domid >>| fun () ->
             let c = Qmp_protocol.connect (monitor_path domid) in
             Qmp_protocol.negotiate c;
+            Qmp_protocol.write c (Command (None, Cont));
```
Comment (mseri) [inline xc/device.ml]: This is to start the CPU right?

**Comment by rosslagerwall** [comment]
@mseri We want to receive the QMP event that QEMU has started execution rather than using xenstore but there is a race between connecting to the QMP event socket and the event being emitted. So QEMU is now started paused and the CONT command is sent to begin execution. Only after that will the event be emitted -- there is no race anymore.

**Comment by mseri** [comment]
Tanks for the explanation

**Pull request merged**